server/base/src/main/java/org/apache/accumulo/server/conf/codec/VersionedPropCodec.java

```diff
@@ -0,0 +1,285 @@
+      new EncodingOptions(dis);
+      return new DataVersionInfo(dis).getDataVersion();
+    } catch (NullPointerException | IOException ex) {
+      throw new IllegalArgumentException(
+          "Failed to read data version version from byte array provided", ex);
+    }
+  }
+
+  /**
+   * Decode the payload and any optional encoding specific metadata and return a map of the property
+   * name, value pairs.
+   *
+   * @param inStream
+   *          an input stream
+   * @param encodingOpts
+   *          the general encoding options.
+   * @return a map of properties name, value pairs.
+   * @throws IOException
+   *           if an exception occurs reading from the input stream.
+   */
+  abstract Map<String,String> decodePayload(final InputStream inStream,
+      final EncodingOptions encodingOpts) throws IOException;
+
+  /**
+   * Read the property map from a data input stream as UTF strings. The input stream should be
+   * created configured by sub-classes for the output of the sub-class. If the sub-class uses an
+   * encoding other that UTF strings, they should override this method. An example would be an
+   * encoding that uses JSON to encode the map.
+   * <p>
+   * The handling the properties as UTF strings is one implementation. Subclasses can implement
+   * different mechanism if desired, one example might be using a JSON implementation to encode /
+   * decode the properties.
+   *
+   * @param dis
+   *          a data input stream
+   * @return the property map
+   * @throws IOException
+   *           if an exception occurs reading from the stream.
+   */
+  Map<String,String> readMapAsUTF(DataInputStream dis) throws IOException {
+
+    Map<String,String> aMap = new HashMap<>();
+    int items = dis.readInt();
+
+    for (int i = 0; i < items; i++) {
+      String k = dis.readUTF();
+      String v = dis.readUTF();
+      aMap.put(k, v);
+    }
+    return aMap;
+  }
+
+  /**
+   * Write the property map to the data output stream. The underlying stream is not closed by this
+   * method.
+   * <p>
+   * The handling the properties as UTF strings is one implementation. Subclasses can implement
+   * different mechanism if desired, one example might be using a JSON implementation to encode /
+   * decode the properties.
+   *
+   * @param dos
+   *          a data output stream
+   * @param aMap
+   *          the property map of k, v string pairs.
+   * @throws IOException
+   *           if an exception occurs.
+   */
+  void writeMapAsUTF(final DataOutputStream dos, final Map<String,String> aMap) throws IOException {
+
+    dos.writeInt(aMap.size());
+
+    for (Map.Entry<String,String> e : aMap.entrySet()) {
+      dos.writeUTF(e.getKey());
+      dos.writeUTF(e.getValue());
+    }
+    dos.flush();
+  }
+
+  /**
+   * Helper class for reading / writing versioned properties metadata.
+   */
+  static class DataVersionInfo {
+    private final int dataVersion;
+    private final Instant timestamp;
+
```

Comment:
It is used - its is a holdover an maybe overkill for just the two values - but I like the delegation so that it does not need to be handled in multiple places.  Will look at using a static method like previous suggestion.